### PR TITLE
Updated normalizeValue

### LINKED ISSUE #74: GraphQL error: String cannot represent value
After updating, I’m consistently getting an error with any embed: “String cannot represent value”

```json
{
  "message": "String cannot represent value: {\"url\":\"https:\\/\\/www.youtube.com\\/watch?v=z9Ug-3qhrwY\",\"oembed\":null}",
  "extensions": {
    "category": "graphql"
  }
}
```

Using Craft 3.6.12.

@reganlawton Do you have any idea what might be causing that?

### PR DESCRIPTION
**Pull request by joshuabaker**
Fixes #74

### CODE REVIEW

**Comment by joshuabaker** [comment]
@reganlawton Please can you review this. 

One key things that I’ve dropped is `$this->oembed`. I couldn’t understand why this was included when it’s absent in the first two returns of your code.

The second `ArrayHelper::getValue` call is there to avoid the error we’re seeing in #74 and should be kept for all sites that might have incorrectly saved the nested value.

File: src/fields/OembedField.php [hunk 1]
```diff
@@ -105,23 +105,34 @@ public function getContentGqlType()
...
+            $value = ArrayHelper::getValue($value, 'url', $url);
+        }
+        
+        // If URL string, return an instance of `OembedModel`
+        if (UrlHelper::isFullUrl($value)) {
+            return new OembedModel($value);
+        }
+        
+        // If we get here, something’s gone wrong
+        return null;
```
Comment (joshuabaker) [inline src/fields/OembedField.php:135]: @reganlawton I did wonder if this should be a throw.

Comment (reganlawton) [reply src/fields/OembedField.php:135]: Adding logging might be a "nice to have" but it can be added later

**Comment by reganlawton** [comment]
Hi @joshuabaker thanks for the PR I'm been / currently recovering from a stroke, so I've been MIA lately and great for your code and only just starting to get back in too things myself.

The `$this->oembed` might have been older code to speed performances and my only thinking will be to ask if your able to change your code works with caching.

If you can change that I'll be happy to merge and what ask that you update the changelog and `composer.json` version. I'd be happy to have you to add a contribution note on README and/ or CHANGELOG file

**Comment by joshuabaker** [comment]
> I'm been / currently recovering from a stroke

Sorry to hear that and wishing you a speedy, full recovery!

> if your able to change your code works with caching

Was the `$this->oembed` doing that specifically? My sense is that caching is a separate concern and should probably be tackled separately; Likely in the service itself. Happy to discuss.

**Comment by denisyilmaz** [comment]
> I'm been / currently recovering from a stroke
same from me! hope you get better soon.

@joshuabaker 
I tested your updated `OembedField.php` in my project that has the issue we were discussing here #74 
It seems it is missing the helpers in the head of the file. After adding these lines it works:

```php
use craft\helpers\Json;
use craft\helpers\ArrayHelper;
use craft\helpers\UrlHelper;
```

My NEO field with oembed is now working with multiple langues thanks to your fix.
__But:__ it need to add the oembed field block again to the entry to solve the issue. existing oembed blocks fail.

**Comment by joshuabaker** [comment]
> It seems it is missing the helpers in the head of the file

That’ll serve me right for just pasting from my local. Updated.

> `Argument 1 passed to craft\\helpers\\UrlHelper::isFullUrl() must be of the type string, array given`

This is an interesting one, because it suggests there’s no value for `url` in the ‘outer’ JSON.

**Comment by denisyilmaz** [comment]
I double checked with cleared caches. This error is happening to the oembed fields that were throwing the error from #74 before (where there seems to be something mis-saved in the database). I'm not to familiar with migrations in craft but I think we must clear the saved values for the oembed field with the update so your new logic is used on every occurance of the oembed field

**Review by denisyilmaz — COMMENTED** [review]
this fixed the error i mentioned in the PR!

**Comment by joshuabaker** [comment]
@denisyilmaz Careful that some aren’t `null`. I’m realising that this is a recursive issue. See the third row of [my comment](https://github.com/wrav/oembed/issues/74#issuecomment-849741242). In that row `url` is nested twice, so there will be select scenarios where even the second check fails (or returns `null` silently). It’s likely we’ll need a `while` loop to check `value` isn’t an array with a `url` value, and then throw after if `value` is still an array.

Broader changes are going to be needed here that this PR. For instance, why are we storing JSON and then just throwing away all but the `url` when we normalise? Lots to clean up.

**N.B.** This PR is to fix #74 alone.

**Comment by denisyilmaz** [comment]
wouldnt this be enough (Line 126):

```php
// If array with `url` attribute, that’s our url so update the value
while(is_array($value)) {
    $value = ArrayHelper::getValue($value, 'url');
}
```

**Comment by denisyilmaz** [comment]
is there a reason thats blocking the merge? @joshuabaker were you able to check the recursive issue? was my `while` loop helping?

**Comment by joshuabaker** [comment]
@denisyilmaz Your approach makes sense and I’m not intending to block the merge here (deep in something else). As noted above, my changes are just to address #74 and not the other things I noted.

@reganlawton asked me to [address caching](https://github.com/wrav/oembed/pull/75#issuecomment-850149771), however, the current implementation doesn’t cache anything, instead throwing away the JSON per session. Not sure how to proceed.

**Comment by denisyilmaz** [comment]
This should take care of caching: 
```php
    /**
     * @var null|mixed
     */
    private $_value;

    /**
     * @param mixed $value The raw field value
     * @param ElementInterface|null $element The element the field is associated with, if there is one
     *
     * @return mixed The prepared field value
     */
    public function normalizeValue($value, ElementInterface $element = null)
    {

        if ($this->_value !== null) {
            return $this->_value;
        }
        $this->_value = null;

        // If null, don’t proceed
        if ($value === null) {
            return null;
        }
        
        // If an instance of `OembedModel` and URL is set, return it
        if ($value instanceof OembedModel && $value->url) {
            $this->_value = $value;
        }

        // If JSON object string, decode it and use that as the value
        if (Json::isJsonObject($value)) {
            $value = Json::decode($value); // Returns an array
            $this->_value = $value;
        }

        // If array with `url` attribute, that’s our url so update the value
        while(is_array($value)) {
            $value = ArrayHelper::getValue($value, 'url');
            $this->_value = $value;
        }
        
        
        // If URL string, return an instance of `OembedModel`
        if (is_string($value) && UrlHelper::isFullUrl($value)) {
            $this->_value = new OembedModel($value);
        }
        
        return $this->_value;
    }
```

**Comment by reganlawton** [comment]
I moved this PR to use the branch `alpha` does anyone have an issue with me pull this and changing the change from @denisyilmaz and then getting a version bump?

**Comment by reganlawton** [comment]
This will make using in composer easier at the very least for testers

**Comment by joshuabaker** [comment]
> I moved this PR to use the branch `alpha`

Wait, why? The alpha branch isn’t production stable, I take it? We should be fixing this for the latest production-ready release. The changes can merge to `alpha` too.

**Comment by reganlawton** [comment]
@joshuabaker  Are you about to merge @denisyilmaz changes or did you want me to make it in another branch as I cant merge your code without merging your branch.

**Comment by joshuabaker** [comment]
@reganlawton No, got for it. Thanks 😊

**Comment by joshuabaker** [comment]
@reganlawton Where are we at with this one? If there’s more to be done, please can you provide a checklist of tasks.

**Comment by joshuabaker** [comment]
> @reganlawton changed the base branch from master to alpha 7 days ago

I just noticed that you went ahead and merged this with the `alpha` branch. This is a hotfix for the `master` branch!

The issue is currently impacting production sites.

**Comment by reganlawton** [comment]
I'm on this today

**Pull request merged**